player/score_bounded_mcts/score_bounded_mcts_node.py

```diff
@@ -0,0 +1,55 @@
+import numpy as np
+
+from game_logic.game_status import GameStatus
+from game_logic.igame_state import IGameState
+from player.mcts.mcts_node import MctsNode
+from player.mcts.imcts_node import IMctsNode
+
+
+class ScoreBoundedMctsNode(MctsNode):
+    def __init__(self, game_state: IGameState, action: int, parent: IMctsNode):
+        super().__init__(game_state, action, parent)
+        self._score_lower_bound = -np.inf
+        self._score_upper_bound = np.inf
+
+    def select(self, node, state):
+        while not node.untried_actions and node.game_state.available_moves:
+            node = node.select_child(node.game_state.is_first_player_move)
+            state.make_move(node.action)
+        return node, state
+    
+    def backpropagate(self, node, state):
+        current_reward = state.get_results_for_player(self.game_state.is_first_player_move)
+
+        node._number_of_runs += 1
+        is_terminal_node = node.state.game_status != GameStatus.InProgress
+        node._reward += current_reward
+        node._reward_squared += current_reward ** 2
+        node.update_score_bounds(current_reward, node.game_state.is_first_player_move, is_terminal_node)
+        node = node.parent
+        
+        while node:
```

Comment:
I guess upgrading node statistics can still be doned in loop. I was thinking of calling `update_score_bounds` for terminal node before loop. Then in loop upgrade statistics and after moving to parent call `update_score_bounds`. Downside is that it will be called also for root, but we can eventually prevent that by explicit check whether parent is not null